NEW TEST [ MacOS ] imported/w3c/web-platform-tests/media-source/dedicated-worker/mediasource-worker-detach-element.html is a flaky crash

jyavenard opened the pull request:
#### 5f3622e3b24527b6a58d4b50c69f1a68e4709eea
<pre>
NEW TEST [ MacOS ] imported/w3c/web-platform-tests/media-source/dedicated-worker/mediasource-worker-detach-element.html is a flaky crash
<a href="https://bugs.webkit.org/show_bug.cgi?id=269871">https://bugs.webkit.org/show_bug.cgi?id=269871</a>
<a href="https://rdar.apple.com/123404489">rdar://123404489</a>

Reviewed by Youenn Fablet.

When shutting down the RemoteMediaSourceProxy in the GPUP we would send
a asynchronous message to the MediaSourcePrivateRemote in the WP which in
turn would have disconnected all the SourceBufferPrivateRemote.
However, it was possible at this stage that the MediaSource in a worker
to itself have started to shutdown and detaching all existing SourceBuffer.
At this stage, the MediaSourcePrivateRemote::m_sourceBuffers would have
been emptied and be unable to disconnect them.

So we split the shutdown of the RemoteMediaSourceProxy and of the
RemoteSourceBufferProxy so that they each inform their respective
remotes to start shutting down.
Similar to what the RemoteMediaSourceProxy does when being shutdown by
the RemoteMediaPlayerProxy, it will now tell all RemoteSourceBufferProxies
to shutdown, which in turn will shutdown the source buffers in the
web process.
Only once that message been sent that it will continue tearing down the
proxies.

Covered by re-enabled test.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::disconnect):
(WebKit::RemoteMediaSourceProxy::shutdown):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::shutdown):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::MessageReceiver::mediaSourcePrivateShuttingDown):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::removedFromMediaSource):
(WebKit::SourceBufferPrivateRemote::MessageReceiver::sourceBufferPrivateShuttingDown):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemoteMessageReceiver.messages.in:

Canonical link: <a href="https://commits.webkit.org/275268@main">https://commits.webkit.org/275268@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de37e421726907b06a81f0f7536ced2db322d6f8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41391 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20405 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43769 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43956 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/37483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23496 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17735 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41965 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/17321 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/35638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14872 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/36641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45300 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/37568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/36961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39107 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/17825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9272 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/17878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
```diff
@@ -330,6 +330,16 @@ void RemoteSourceBufferProxy::setMaximumQueueDepthForTrackID(TrackID trackID, ui
+    if (!m_connectionToWebProcess)
+        return;
+
+    m_connectionToWebProcess->connection().sendWithAsyncReply(Messages::SourceBufferPrivateRemoteMessageReceiver::SourceBufferPrivateShuttingDown(), [this, protectedThis = Ref { *this }, protectedConnection = Ref { *m_connectionToWebProcess }] {
```
Theoretically, RefPtr for m_connectionToWebProcess

jyavenard replied:
a RefPtr is taken!

youennf commented on Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
```diff
@@ -149,8 +148,9 @@ class SourceBufferPrivateRemote final
-    bool isGPURunning() const { return !m_disconnected; }
-    bool m_disconnected { false };
+    bool isGPURunning() const { return !m_shutdown && !m_removed; }
+    std::atomic<bool> m_shutdown { false };
```
m_isShuttingDown?

jyavenard replied:
the GPU Process is shutting down, once this method is finished, it would be shutdown.

so m_isShutdown would be more suitable; however `m_shutdown` is more used in the tree than `m_isShutdown` and is what was used in MediaSourcePrivateRemote

youennf commented on Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
```diff
@@ -149,8 +148,9 @@ class SourceBufferPrivateRemote final
-    bool m_disconnected { false };
+    bool isGPURunning() const { return !m_shutdown && !m_removed; }
+    std::atomic<bool> m_shutdown { false };
+    bool m_removed { false };
```
m_isRemoved?
Also, I am not exactly sure why m_shutdown is atomic but not m_removed.
They seem to be accessed from both main thread (callOnMainThreadAndWait) and work queue

jyavenard replied:
m_removed is only accessed in the dispatcher; m_shutdown is immediately accessed in the receiver without a redispatch. but for consistency I'll make it atomic too

jyavenard committed:
NEW TEST [ MacOS ] imported/w3c/web-platform-tests/media-source/dedicated-worker/mediasource-worker-detach-element.html is a flaky crash
https://bugs.webkit.org/show_bug.cgi?id=269871
rdar://123404489

Reviewed by Youenn Fablet.

When shutting down the RemoteMediaSourceProxy in the GPUP we would send
a asynchronous message to the MediaSourcePrivateRemote in the WP which in
turn would have disconnected all the SourceBufferPrivateRemote.
However, it was possible at this stage that the MediaSource in a worker
to itself have started to shutdown and detaching all existing SourceBuffer.
At this stage, the MediaSourcePrivateRemote::m_sourceBuffers would have
been emptied and be unable to disconnect them.

So we split the shutdown of the RemoteMediaSourceProxy and of the
RemoteSourceBufferProxy so that they each inform their respective
remotes to start shutting down.
Similar to what the RemoteMediaSourceProxy does when being shutdown by
the RemoteMediaPlayerProxy, it will now tell all RemoteSourceBufferProxies
to shutdown, which in turn will shutdown the source buffers in the
web process.
Only once that message been sent that it will continue tearing down the
proxies.

Covered by re-enabled test.
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebKit/GPUProcess/media/RemoteMediaSourceProxy.cpp:
(WebKit::RemoteMediaSourceProxy::disconnect):
(WebKit::RemoteMediaSourceProxy::shutdown):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::shutdown):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/MediaSourcePrivateRemote.cpp:
(WebKit::MediaSourcePrivateRemote::MessageReceiver::mediaSourcePrivateShuttingDown):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::removedFromMediaSource):
(WebKit::SourceBufferPrivateRemote::MessageReceiver::sourceBufferPrivateShuttingDown):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemoteMessageReceiver.messages.in:

Canonical link: https://commits.webkit.org/275268@main
